Engine/GameEngine.cs

```diff
@@ -0,0 +1,410 @@
+    public const int MaxHandCardsBase = 5;
+    public const int CardsInKitty = 15;
+    public const int CardsInTopUp = 5;
+
+    public static GameState NewGame(List<string>? playerNames = null,
+        Settings? settings = null)
+    {
+        // Initialise player names if none were supplied
+        if (playerNames == null)
+        {
+            playerNames = new List<string>();
+            for (var i = 0; i < PlayersPerGame; i++) playerNames.Add($"Player {i + 1}");
+        }
+
+        // Create deck
+        var deck = new List<Card>();
+        for (var i = 0; i < CardsInDeck; i++)
+        {
+            var suit = (Suit) (i / CardsPerSuit);
+            int value = i % CardsPerSuit;
+            var newCard = new Card
+            {
+                Id = i,
+                Suit = suit,
+                Value = value
+            };
+            deck.Add(newCard);
+        }
+
+        // Shuffle the new deck
+        deck.Shuffle(settings?.RandomSeed);
+
+        // Deal cards to players
+        var players = new List<Player>();
+        for (var i = 0; i < PlayersPerGame; i++)
+        {
+            players.Add(new Player
+            {
+                Id = i + 1,
+                Name = playerNames?[i] ?? $"Player {i + 1}",
+                HandCards = deck.PopRange(MaxHandCardsBase),
+                KittyCards = deck.PopRange(CardsInKitty),
+                TopUpCards = deck.PopRange(CardsInTopUp)
+            });
+        }
+
+        var gameState = new GameState
+        {
+            Players = players,
+            CenterPiles = new List<List<Card>> {deck.PopRange(1), deck.PopRange(1)},
+            Settings = settings
+        };
+        return gameState;
+    }
+
+    public static Result<Player> TryGetWinner(GameState gameState)
+    {
+        // Check if a player has no cards in their hand or kitty
+        foreach (Player player in gameState.Players)
+        {
+            if (player.HandCards.Count <= 0 && player.KittyCards.Count <= 0)
+            {
+                // Return any player that does
+                return new SuccessResult<Player>(player);
+            }
+        }
+
+        return new ErrorResult<Player>("No winner yet!");
+    }
+
+    public static Result AllPlayersRequestingTopUp(GameState gameState)
+    {
+        // Check all players are requesting top up
+        return !gameState.Players.All(p => p.RequestingTopUp)
+            ? new ErrorResult("Not all players are requesting top up")
+            : new SuccessResult();
+    }
+
+    public static Result<GameState> TryTopUp(GameState gameState)
+    {
+        // Check all players are requesting top up
+        Result canTopUpResult = CanTopUp(gameState);
+        if (canTopUpResult is IErrorResult canTopUpError)
+        {
+            return new ErrorResult<GameState>(canTopUpError.Message);
+        }
+
+        // Make sure we have cards in our top up pile
+        if (gameState.Players[0].TopUpCards.Count <= 0)
```

Comment:
I dont actually know how this game works but a cheeky 0 index here looks suspect. Is there not some sort of player turn variable inside gamestate?

Reply:
For this game everyone always has the same amount of cards in their top up piles. Maybe I could create a TopUpPileCardCount in GameState instead?